re-lint only changed rules

tophf opened the pull request:
Since both our linters are slow on LARGE code sections, this PR implements partial linting of the changed text. Each change is expanded to its encompassing rule via dumb `}` matching and is extracted along with 2-3 adjacent rules. This extracted fragment is passed to the linter. The new results are merged with the old ones that were outside of the newly checked ranges.

The main advantage is that the time spent will be almost the same regardless of a code section size. According to my perf tests on a fast computer, lint-on-typing consumes just a few ms per change. Now the bottleneck is CodeMirror itself.

I've temporarily changed the delay to update lint report panel to 500ms to demonstrate the increased responsiveness.

[relint-changes.zip](https://github.com/openstyles/stylus/archive/relint-changes.zip).

narcolepticinsomniac commented:
I haven't really been testing the linting options, but you guys have done a great job with them. It's nice that they're warnings now, but I'd still strongly recommend this default:

```
    "selector-type-no-unknown": [
      false, {"severity": "warning"}
    ],
```
The CM code folding arrows in the gutter seem slightly bugged. They're showing up real late in the page load.

I also got this error:

![cm](https://user-images.githubusercontent.com/14100003/29795508-8a04591a-8c1a-11e7-9cdf-6c4842937737.png)


As for "re-linting only changed rules", I'm sure you can analyze the numbers performance-wise, and I imagine they're significant. All I can say is that both linters seem very responsive, whereas Stylelint was slightly laggier when I last tested, and it's not like CSSLint was ever flawless either. I'm guessing it's a combo of refining the enforced rules and "re-linting only changed rules", but they both seem more responsive.

Mottie commented:
I'm seeing the same error that @narcolepticinsomniac reported. It occurs after you cancel out of the stylelint config, then reopen the config.

I also added a log to test the code

```js
  function stylelintOnRange(range) {
    console.log(range);
```

I noticed the following:

* On init, I see the log twice - I mentioned this before and I think it's CodeMirror calling the linter twice
* Opening stylelint's config, the linter is called once on an empty string.
* Clicking on "Save" the linter is run twice.
* I modify the config (set the "selector-type-no-unknown" to `false` on the GitHub Dark style) then click save, it runs the linter on the entire style four times.
   ![](https://user-images.githubusercontent.com/136959/29798622-b1e480b4-8c23-11e7-80a9-91ffbc259263.png)

narcolepticinsomniac commented:
Another related error:

![cm-2](https://user-images.githubusercontent.com/14100003/29798841-8e674a4a-8c2d-11e7-9e8e-ca396000ea52.png)

Mottie commented:
One more... when opening the editor and clicking on "beautify" with stylelint set as the linter

> Uncaught ReferenceError: exports is not defined
    at setTimeout (edit.js:1135)

Also, after beautifying (after switching to CSSLint and back to Stylelint), the "update" is only seeing the first block of style and end up removing all lint errors. The issue block remains blank even after I click on Undo.

![](https://user-images.githubusercontent.com/136959/29798922-bae8b7d2-8c25-11e7-9d54-a6afa474ddf8.png)

Mottie commented:
I had one other concern with only linting a portion of the style that includes variables; but it appears that Stylelint doesn't have an issue with undefined variables (yet) - enter `a { color: var(--undef); }` in [their demo site](https://stylelint.io/demo/) and there is no error.

----

Unrelated to this PR, CSSLint is having issues with variables added along with other properties (try [this style](https://userstyles.org/styles/137280/medium-dark-grey)):

```css
.twitter-card {
    border-bottom: 1px solid var(--color_bg4);
}
```

> Expected end of value but found 'var(--color_bg4)'.

narcolepticinsomniac commented:
I've been trying out Stylelint. I haven't done a lot of editing so far, but I was investigating a redundant code warning, which was correct, but when I fixed it there were a lot of issues with the linter not adjusting to changes, and some weird incorrect errors I think were related to the code being in a screen. I made a screen record. 

[Stylelint.zip](https://github.com/openstyles/stylus/files/1259702/Stylelint.zip)

Maybe it's only having issues because of the screen. It throws an error on paste, removing redundant code doesn't remove warnings and actually throws an extra error till saved and then they all disappear. Deleting lines and leaving valid code also throws errors, but I'm pretty sure that's because it's not dealing with the screen well.

Here's the code:

```
@media (min-width: 177.778vh) {
.post_content iframe:nth-of-type(2) {
    margin-top:100vh!important;
}
.post_content iframe:nth-of-type(3) {
    margin-top: 200vh!important;
}
.post_content iframe:last-of-type {
    margin-bottom: 30px!important;
}
.post_content iframe {
    margin-left: auto!important;
    margin-right: auto!important;
    top: 22px!important;
    position: absolute!important;
    left: 0!important;
    right: 0!important;
    height: calc(100vh - 30px)!important;
    min-height: calc(100vh - 30px)!important;
    max-height: calc(100vh - 30px)!important;
    width: calc(((100vh - 30px) * 16) / 9)!important;
    min-width: calc(((100vh - 30px) * 16) / 9)!important;
    max-width: calc(((100vh - 30px) * 16) / 9)!important;
    padding: 0!important;
    float: none!important;
    clear: none!important;
    margin: 0 auto!important;
}
}
```

narcolepticinsomniac commented:
Btw, I do realize all the positioning code at the bottom isn't necessary with the absolute positioning code at the top. The video sizing code is a macro, but in this particular instance I added the absolute code for the site and didn't bother deleting irrelevant code.

tophf commented:
Thank you for the details and examples! I think I've fixed most of the issues so you can test more. There are still some edge cases to handle, but I'll do it tomorrow.

> I'd still strongly recommend this default:    "selector-type-no-unknown": [

AFAICT most styles use standard DOM names so having this option by default would be helpful. It can be configured anyway. Later we'll add a rightclick or a button in `Issues (i)` report to silence any rule.

> The CM code folding arrows in the gutter seem slightly bugged. They're showing up real late in the page load.

The only difference I saw and fixed is that the linting marks gutter was added a bit later so it shifted the code (introduced in 29851a8de9a2402308cc4ee60630156a07a55373 which enabled lazy-loading of linters).

narcolepticinsomniac commented:
Much better! Only issue I still see with the example code is when "write a new style" is selected, the linter doesn't kick in on the initial paste.

>...linting marks ...lazy-loading

That explains it. They're still showing up a bit late, but I guess that's intended behavior now. The code is shifting is fixed, which was the real issue.

>AFAICT most styles use standard DOM names so having this option by default would be helpful. 

Sure, most stick to standard selectors, but odd selectors which throw warnings aren't all that uncommon. NBD though, maybe it bugs me more than it should.

> Later we'll add a rightclick or a button in Issues (i) report to silence any rule.

That'd be awesome. I imagine some casual users may be intimidated editing a JSON for preferences. What would be perfect is if there could be two options for silencing, one to silence the rule altogether and one to silence the rule for that particular warning. 

For example:

Unknown pseudo-class selector ":decrement"

The incorrect pseudo-class warnings on ":increment" and ":decrement" I'd like to silence without disabling the check on pseudo-classes altogether.

Also, checking the scrollbar style, those two warnings lit up like a Christmas tree before, but now it's showing only one for each. When you remove that instance, the warning moves on to the next instance. Not sure if that's a bug or a feature. On one hand, it's much cleaner, but OTOH, it's misleading as to how many actual issues the code has.

Code: https://pastebin.com/a2ptqTW6

tophf commented:
> now it's showing only one for each

It was a bug in my code, now fixed. As well as the pasting bug.

Mottie commented:
Do you think it would be easier to present a form to the user with each rule (with a link) plus a toggle and severity dropdown? Then we wouldn't have to worry about them dealing with JSON.

tophf commented:
> present a form to the user with each rule (with a link) plus a toggle and severity dropdown

Actually, this is how I thought the config UI would look like, but then I don't know how complex a stylelint config json may be. Doesn't it allow adding additional tweaks to each rule besides the common on/off and severity? If yes, it'd be tricky to implement, but still possible I guess.

Mottie commented:
Found a minor issue with the config...

The `"no-empty-source": false` is not wrapped in an array, but when I open it in the config I see:

```js
    "no-empty-source": [
      false
    ]
```

Here is a screenshot of the stylelint results object:

![](https://user-images.githubusercontent.com/136959/29843080-67fe3654-8cd0-11e7-928c-cd2ec067a12c.png)

Mottie commented:
>  Doesn't it allow adding additional tweaks to each rule besides the common on/off and severity? If yes, it'd be tricky to implement, but still possible I guess.

Yes, like the `"declaration-block-no-duplicate-properties"` rule... JSON would be the source of the code that builds the form, so it *shouldn't be too difficult* (famous last words). I'll take a look when I get some time later this evening.

narcolepticinsomniac commented:
>As well as the pasting bug.

Fixed for "write a new style", but similarly bugged for Moz-imported code. Also, after importing, there is no browser alert for attempting to reload or close the editor page without saving.

narcolepticinsomniac commented:
>Or maybe we're overthinking it... if they know how to edit CSS, they probably know how to edit JSON.

Perhaps. I tend to find it annoying how most devs presume the ineptitude of the end-user and implement features accordingly, especially when slightly more complicated configurations give the user more options. We are all guilty to some degree, and for good reason. However inept you presume the end-user to be, reviews and bug reports will confirm that you have underestimated.

Mottie commented:
Now it seems that the re-linted code block starts from where the edit occurred and goes to the end of the file

Caused an error on line 44

![](https://user-images.githubusercontent.com/136959/29844894-744d3a12-8cd6-11e7-8e4b-c650078ae33b.png)

Caused an error on line 2112

![](https://user-images.githubusercontent.com/136959/29844910-8528e246-8cd6-11e7-8acd-74ec35c00f23.png)

narcolepticinsomniac commented:
Browser alert for saving is fixed, but Moz-imported code still isn't validated on import.

Mottie commented:
YAY another fun issue...

I make a minor edit (adding a space) to a style then saved it. Then I opened the manager page in another tab and clicked on "Check for update", then installed the original style. Going back to the original tab I found this:

![](https://user-images.githubusercontent.com/136959/29846831-2c8dde54-8cde-11e7-8fb0-47525cc219f8.png)

The "Code 2" block popped up and was clickable whereas the link in "Code 1" was did nothing on click.

I'm sure you're loving us now @tophf.

tophf commented:
Yeah I love the easily reproducible stuff. Fixed.

tophf commented:
Attempting to paste moz-format userstyle in any section editor now shows the import dialog with the pasted text.

narcolepticinsomniac commented:
>Attempting to paste moz-format userstyle in any section editor now shows the import dialog with the pasted text.

That's pretty great, and appears to be working well. Not to mention that errors for declining to save before unloading the editor are now gone, which is a nice bonus. 

Beautifier seems fixed with either linter, and validator kicks in after Moz-import. Functionality is vastly improved all around. Well done!

Mottie commented:
When the re-linter fires, it looks like the `code` that is passed to the linter is correct, but the `from` and `to` line number values are one less than they what they should be. The `ch` is correct.

After Moz-import, I noticed that the validator is using CSSLint instead of my selected Stylelint. Is this intended?

Otherwise everything seems pretty solid now.

tophf commented:
> When the re-linter fires, it looks like the code that is passed to the linter is correct, but the from and to line number values are one less than they what they should be. The ch is correct.

linter uses 1-based numbers but my code is 0-based just like CodeMirror so I guess this is the difference you see. The `ch` is "correct" because it points at the end of line after `}`.



> After Moz-import, I noticed that the validator is using CSSLint instead of my selected Stylelint. Is this intended?

Sounds like a bug but I couldn't reproduce.

Mottie commented:
> Sounds like a bug but I couldn't reproduce.

Oops, I should have included an example... Set the linter to "stylelint", then import and append this snippet:

```css
body {
  background-color: rgba(/*[[base-color-rgb]]*/, .2) !important;
}
```

I know it isn't valid, but it is usually combined with a valid bg color.

Once appended, I see this issue pop up: "2:52 Expected RBRACE".

tophf commented:
Ah, that small report is produced by the parser which is a part of CSSLint indeed so this is working as intended.

narcolepticinsomniac commented:
A couple issues in CSSLint. It's throwing warnings for browser specific gradients when it didn't used to, and it's throwing an incorrect error when editing code directly after commented out code. 

Screen record:

[CSSLint.zip](https://github.com/openstyles/stylus/files/1265486/CSSLint.zip)

Code:

```
/* button {
    text-shadow: none!important;
    background: none!important;
    box-shadow: inset 0 2px 0 0 hsl(0, 0%, 22%), inset 0 7px 0 3px hsl(0, 0%, 24%), inset 0 -2px 0 0 hsl(0, 0%, 18%), 0 1px 6px hsla(0, 0%, 0%, .15), 0 1px 4px hsla(0, 0%, 0%, .3)!important;
    padding: 2px 6px!important;
    background-color: hsl(0, 0%, 12%)!important;
    color: hsl(0, 0%, 90%)!important;
    border: 1px solid hsl(0, 0%, 38%)!important;
    border-radius: 1px!important;
    line-height: unset!important;
    transition: background-color .2s ease, border-color .2s ease, box-shadow .2s ease, color .2s ease!important;
}
button:hover {
    box-shadow: inset 0 2px 0 0 hsl(0, 0%, 28%), inset 0 7px 0 3px hsl(0, 0%, 30%), inset 0 -2px 0 0 hsl(0, 0%, 24%), 0 1px 8px hsla(0, 0%, 0%, .3), 0 1px 6px hsla(0, 0%, 0%, .6)!important;
    background-color: hsl(0, 0%, 16%)!important;
    color: hsl(0, 0%, 96%)!important;
    border-color: hsl(0, 0%, 44%)!important;
} */
button {
    background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAIAAAAoCAQAAAB+iOP/AAAAAmJLR0QA/4ePzL8AAAAJcEhZcwAACxMAAAsTAQCanBgAAAAHdElNRQfhCBkTHDNLahAxAAAAHWlUWHRDb21tZW50AAAAAABDcmVhdGVkIHdpdGggR0lNUGQuZQcAAABhSURBVAjXVcpLCgIxAATRl9gITmblxlt4/2t4B28g+Blh4sIh6Kaopqv0M1mR7s/GrKuu0mVBFmWAvMfxQp7IfeCBXLCrp6vcdNkjR6QpMiEHNmu6NGQemBSZ9c3at/uNP/cTHeX/1YOFAAAAAElFTkSuQmCC)!important;
    background-color: hsl(0, 0%, 9%)!important;
    background-repeat: repeat-x!important;
    background-size: auto 100%!important;
    background-position: center!important;
    border: 1px solid hsl(0, 0%, 42%)!important;
    border-radius: 1px!important;
    box-shadow: 0 0 0 1px #000!important;
    color: hsl(0, 0%, 86%)!important;
    font: 13.3333px Arial!important;
    min-height: unset!important;
    min-width: unset!important;
    padding: 1px 6px!important;
    text-shadow: none!important;
    transition: background-image .3s, background-color .3s, border-color .3s, color .3s!important;
}
button:hover {
    background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAIAAAAoCAQAAAB+iOP/AAAAAmJLR0QA/4ePzL8AAAAJcEhZcwAACxMAAAsTAQCanBgAAAAHdElNRQfhCBkUGBZhTRf3AAAAHWlUWHRDb21tZW50AAAAAABDcmVhdGVkIHdpdGggR0lNUGQuZQcAAABhSURBVAjXVcpLCgIxAATRl9gITmbl1nN4/zN4B28g+Blh4sIh6Kaopqv0M1mR7s/GrKuu0mVBFmWAvMfxQp7IfeCBXLCrp6vcdNkjR6QpMiEHNmu6NGQemBSZ9c3at/uNP/3xHfDcDyceAAAAAElFTkSuQmCC)!important;
    background-color: hsl(0, 0%, 14%)!important;
    border-color: hsl(0, 0%, 52%)!important;
    color: hsl(0, 0%, 96%)!important;
}
#test {
    background-image: -moz-linear-gradient(top, hsla(190, 40%, 70%, .06), hsla(190, 40%, 70%, .09), hsla(190, 40%, 70%, .11), hsla(190, 40%, 70%, .14))!important;
    background-image: -webkit-linear-gradient(top, hsla(190, 40%, 70%, .06), hsla(190, 40%, 70%, .09), hsla(190, 40%, 70%, .11), hsla(190, 40%, 70%, .14))!important;
}
```

tophf commented:
I've been rewriting range detection completely as the current implementation is too dumb.

Mottie commented:
> I've been rewriting range detection

Can you use the [CodeMirror match brackets addon](https://codemirror.net/doc/manual.html#addon_matchbrackets)?

tophf commented:
Maybe. I'm not there yet. I was referring to the detection of changed ranges.

tophf commented:
I've been hacking at this for several days nonstop and wow this is so convoluted in the end, too many edge cases that can break any non-parser-based approach. The matchbrackets addon is also dumb, it can't exclude comments and brackets inside `content` text, for example.

I've got an idea though: I can intercept `parserlib.css.Parser.prototype.fire` in CSSLint and build a structured list of all rules positions during initial linting pass on the entire section so that I can use the list later to know which rules are affected by an editing change.  It'll work only for typing and simple editing, but that's the intended goal anyway. Hopefully, stylelint can be similarly intercepted. I'll look into it later.

Mottie commented:
That sounds like a lot of work when I could disable the linter if I feel it's causing too much lag, then re-enable it when I'm done.

Maybe it'd be easier to keep the simple bracket matching algorithm, but add a refresh the linter button that lints the entire style?

tophf commented:
Maybe, but it's too flaky to rely upon without at least trying to hook the parser first.

tophf commented:
The functionality was implemented differently: our CSSLint was rewritten and its parserlib has a cache now. It provides 10x speedup when CSSLint is the active linter, and for the parsing stage of usercss regardless of the active linter, resulting in the same 10x speedup during continuous config dialog tweaking and live-preview.
